### PR TITLE
#1341 added several locks to p2p_plugin.cpp

### PR DESCRIPTION
**Pull request by NateBrune**
Added several read locks per issue #1341. The plugin should acquire a read lock before fetching database members.

### CODE REVIEW

File: libraries/plugins/p2p/p2p_plugin.cpp [hunk 1]
```diff
@@ -227,7 +227,10 @@ void p2p_plugin::plugin_startup()
 
       my->node->listen_to_p2p_network();
       my->node->connect_to_p2p_network();
-      my->node->sync_from(graphene::net::item_id(graphene::net::block_message_type, my->chain.db().head_block_id()), std::vector<uint32_t>());
+      my->chain.db().with_read_lock( [&]()
+      {
+         my->node->sync_from(graphene::net::item_id(graphene::net::block_message_type, my->chain.db().head_block_id()), std::vector<uint32_t>());
```
Comment (mvandeberg) [inline libraries/plugins/p2p/p2p_plugin.cpp]: Grab head block id in the lock, then call `sync_from`

File: libraries/plugins/p2p/p2p_plugin.cpp [hunk 2]
```diff
@@ -354,55 +361,60 @@ void p2p_plugin_impl::handle_message( const graphene::net::message& message_to_p
...
-   if( blockchain_synopsis.empty()
-         || ( blockchain_synopsis.size() == 1 && blockchain_synopsis[0] == block_id_type() ) )
-   {
-      // peer has sent us an empty synopsis meaning they have no blocks.
-      // A bug in old versions would cause them to send a synopsis containing block 000000000
-      // when they had an empty blockchain, so pretend they sent the right thing here.
-      // do nothing, leave last_known_block_id set to zero
-   }
-   else
+{  chain.db().with_read_lock( [&]()
```
Comment (mvandeberg) [inline libraries/plugins/p2p/p2p_plugin.cpp]: This needs to return the vector. The lambda returns the vector to the caller (with_read_lock) but `get_block_ids` does not return anything. gcc and clang should both have a warning on this method.

File: libraries/plugins/p2p/p2p_plugin.cpp [hunk 3]
```diff
@@ -420,7 +432,10 @@ graphene::net::message p2p_plugin_impl::get_item( const graphene::net::item_id&
       });
    }
 
-   return trx_message( chain.db().get_recent_transaction( id.item_hash ) );
+    chain.db().with_read_lock( [&]()
```
Comment (mvandeberg) [inline libraries/plugins/p2p/p2p_plugin.cpp]: Needs to return

File: libraries/plugins/p2p/p2p_plugin.cpp [hunk 4]
```diff
@@ -260,22 +263,26 @@ void p2p_plugin::set_block_production( bool producing_blocks )
 ////////////////////////////// Begin node_delegate Implementation //////////////////////////////
 bool p2p_plugin_impl::has_item( const graphene::net::item_id& id )
 {
-   try
+   chain.db().with_read_lock( [&]()
```
Comment (mvandeberg) [inline libraries/plugins/p2p/p2p_plugin.cpp]: Needs to return

**Comment by NateBrune** [comment]
Fixed locally, #1503 is blocking to test and re-PR

**Comment by mvandeberg** [comment]
Looks good to me

**Pull request merged**